### PR TITLE
Implemention of inspection #KT-15537

### PR DESCRIPTION
**Pull request by jan-gerling**
This inspection reports usages of IntRange IntRange and endInclusive. These properties can be replaced with equivalent kotlin.collections first and last.
Example: range.start() can be replaced with range.first()

**Issue:**
https://youtrack.jetbrains.com/issue/KT-15537

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+
+class ReplaceIntRangeStartEndInclusiveWithFirstLastInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return dotQualifiedExpressionVisitor { expression ->
+            val fName: FqName? = expression.GetFunctionName()
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt]: `val fName = expression.GetFunctionName() ?: return`

and then remove the null checks down here

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                    expression,
+                    "Could be replaced with `last`",
+                    ReplaceIntRangeEndInclusiveWithLastQuickFix()
+                )
+            }
+        }
+    }
+}
+
+private fun KtDotQualifiedExpression.GetFunctionName(): FqName? {
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt]: should be `getFunctionName`

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    val context = this.analyze()
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt:42]: Maybe `this.analyze(BodyResolveMode.PARTIAL)`? It's cheaper, but I'm not certain about the difference.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+}
+
+class ReplaceIntRangeEndInclusiveWithLastQuickFix : LocalQuickFix {
+    override fun getName() = "Replace 'endInclusive' with 'last'"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val element = descriptor.psiElement as KtDotQualifiedExpression
```
Comment (devbridie) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt:72]: remove `!!` under this and add `?: return` here

**Review by devbridie — CHANGES_REQUESTED** [review]

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt (hunk 3 above)
Comment (jan-gerling) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt:42]: I can't find a proper documentation for the anlyze function, thus I would leave as it is.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt (hunk 1 above)
Comment (jan-gerling) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt]: implemented

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt (hunk 2 above)
Comment (jan-gerling) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt]: implemented

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt (hunk 3 above)
Comment (jan-gerling) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt:42]: rejected change

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt (hunk 4 above)
Comment (jan-gerling) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt:72]: implemented

**Review by devbridie — APPROVED** [review]

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        element.selectorExpression.replace(KtPsiFactory(element).createExpression("first"))
+    }
+}
+
+class ReplaceIntRangeEndInclusiveWithLastQuickFix : LocalQuickFix {
+    override fun getName() = "Replace 'endInclusive' with 'last'"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
```
Comment (abansagi) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceIntRangeStartEndInclusiveWithFirstLastInspection.kt:71]: Dereck told me to remove whitelines in my code, so maybe need to remove these here too 😛

**Review by KilianCallebaut — APPROVED** [review]
looks good!

**Comment by devbridie** [comment]
I rebased to current master to solve merge conflicts. Submitting PR now.

**Comment by jan-gerling** [comment]
Closed, because it was opened on jetbrains/kotlin.

**Pull request closed without merging**